File: Y2/POO/lab6/ex1.cpp
#include <iostream>
#include <string>

using namespace std;


class City{
    protected:
        string name;
        int countyCode;

        City(string name, int countyCode): name(name), countyCode(countyCode){}
        virtual string getName() = 0;
        virtual int getCountyCode() = 0;
        virtual string getType() = 0;
        
        virtual ~City(){cout<<"DESTRUCTOR BAZA\n";};
};

class Rural: public City{

    public:
    Rural(string name = "N/A", int countyCode = 0): City(name,countyCode){}
    string getName(){
        return this->name;
    }

    int getCountyCode(){
        return this->countyCode;
    }

    string getType(){
        cout<<"Im a rural city\n";
        return "rural";
    }
    
};

class Industrial: public City{
public:
    Industrial(string name = "N/A", int countyCode = 0): City(name,countyCode){}
    string getName(){
        return this->name;
    }

    int getCountyCode(){
        return this->countyCode;
    }

    string getType(){
        cout<<"Im an industrial city\n";
        return "industrial";
    }

};

class Port: public City{
public:
    Port(string name = "N/A", int countyCode = 0): City(name,countyCode){}
    string getName(){
        return this->name;
    }

    int getCountyCode(){
        return this->countyCode;
    }

    string getType(){
        cout<<"Im a port city\n";
        return "port";
    }
};

int main(){

    Rural orasRural("Las Fierbinti",50);
    cout<<orasRural.getName()<<endl;
    cout<<orasRural.getCountyCode()<<endl;
    orasRural.getType();

    Industrial orasIndustrial("Londra",12);
    cout<<orasIndustrial.getName()<<endl;
    cout<<orasIndustrial.getCountyCode()<<endl;
    orasIndustrial.getType();

    Port orasPort("Mangalia",67);
    cout<<orasPort.getName()<<endl;
    cout<<orasPort.getCountyCode()<<endl;
    orasPort.getType();

}

File: Y2/POO/lab6/kotlin/ex1.kt
abstract class City(
    open var name: String,
    open var countyCode: Int
){
    abstract fun gtName(): String

    abstract fun gtCountyCode(): Int

    abstract fun gtType(): String
}

class Rural(override var name: String = "N/A",
override var countyCode: Int = 0): City(name,countyCode){
    override fun gtName(): String{
        return name
    }

    override fun gtCountyCode(): Int{
        return countyCode
    }

    override fun gtType(): String{
        print("Im a rural city\n")
        return "rural"
    }
}

class Industrial(override var name: String = "N/A",
override var countyCode: Int = 0): City(name,countyCode){
override fun gtName(): String{
        return name
    }

    override fun gtCountyCode(): Int{
        return countyCode
    }

    override fun gtType(): String{
        print("Im an industrial city\n")
        return "industrial"
    }
}

class Port(override var name: String = "N/A",
override var countyCode: Int = 0): City(name,countyCode){
    override fun gtName(): String{
        return name
    }

    override fun gtCountyCode(): Int{
        return countyCode
    }

    override fun gtType(): String{
        print("Im a port city\n")
        return "port"
    }
}
fun main(){
    val orasRural = Rural("Las Fierbinti", 50)
    println(orasRural.gtName())
    println(orasRural.gtCountyCode())
    orasRural.gtType()

    val orasIndustrial = Industrial("Londra", 12)
    println(orasIndustrial.gtName())
    println(orasIndustrial.gtCountyCode())
    orasIndustrial.gtType()

    val orasPort = Port("Mangalia", 67)
    println(orasPort.gtName())
    println(orasPort.gtCountyCode())
    orasPort.gtType()
}